fedimint-api/src/db/mod.rs

```diff
@@ -71,22 +73,73 @@ pub struct Database(Arc<DatabaseInner<dyn IDatabase>>);
 // because it will impossible to construct otherwise
 #[derive(Debug)]
 struct DatabaseInner<Db: IDatabase + ?Sized> {
+    notifications: Notifications,
     module_decoders: ModuleDecoderRegistry,
     db: Db,
 }
 
+const WAIT_KEY_TIMEOUT: Duration = Duration::from_secs(60);
+
 impl Database {
     pub fn new(db: impl IDatabase + 'static, module_decoders: ModuleDecoderRegistry) -> Self {
         let inner = DatabaseInner {
             db,
+            notifications: Notifications::new(),
             module_decoders,
         };
 
         Self(Arc::new(inner))
     }
 
     pub async fn begin_transaction(&self) -> DatabaseTransaction {
-        DatabaseTransaction::new(self.0.db.begin_transaction().await, &self.0.module_decoders)
+        DatabaseTransaction::new(
+            self.0.db.begin_transaction().await,
+            &self.0.module_decoders,
+            &self.0.notifications,
+        )
+    }
+
+    async fn wait_key_inner(&self, key_bytes: &[u8]) -> Result<Vec<u8>> {
+        loop {
+            // register for notification
+            let notify = self.0.notifications.wait(&key_bytes);
+
+            // check for value in db
+            let mut tx = self.0.db.begin_transaction().await;
+            if let Some(value_bytes) = tx.raw_get_bytes(key_bytes).await? {
+                return Ok(value_bytes);
+            } else {
+                // key not found, try again
+                notify.await;
+                // if miss a notification between await and next register, it is fine.
+                // because we are going check the database
+            };
+        }
+    }
+
+    /// Waits for key to be notified.
+    ///
+    /// Returns None if no notify is sent after [`WAIT_KEY_TIMEOUT`].
+    pub async fn wait_key<K>(&self, key: &K) -> Result<Option<K::Value>>
+    where
+        K: DatabaseKey + DatabaseKeyPrefixConst,
+    {
+        let key_bytes = key.to_bytes();
+        let value_bytes =
+            match crate::task::timeout(WAIT_KEY_TIMEOUT, self.wait_key_inner(&key_bytes)).await {
```

Comment:
We don't want a general, global timeout. We could make the timeout an argument, or preferably just let the user deal with it.